### PR TITLE
Location entity codes are only valid along with the parent location entity

### PR DESCRIPTION
**Pull request by frankhuster**
Except for the top-level State entity.
To search for a HealthBlock, for example, you'll need to provide the parent Taluka in addition to the health block code.

### CODE REVIEW

File: region/src/main/java/org/motechproject/nms/region/domain/HealthBlock.java [hunk 1]
```diff
@@ -38,11 +40,14 @@
...
 
     @Field
-    @Unique
     @Column(allowsNull = "false")
     @NotNull
     private Long code;
 
+    // used by CSV import as a placeholder for the taluka
+    @Ignore
+    private String talukaCode;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/domain/HealthBlock.java]: I would avoid adding this kind of fields to entity classes (looks like a code smell). If such a placeholder is really needed, consider creating an intermediate class or use a map to store entity properties.

File: region/src/main/java/org/motechproject/nms/region/domain/HealthFacility.java [hunk 2]
```diff
@@ -40,6 +41,10 @@
     @NotNull
     private HealthFacilityType healthFacilityType;
 
+    // used by CSV import as a placeholder for the healthBlock
+    @Ignore
+    private long healthBlockCode;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/domain/HealthFacility.java]: As above

File: region/src/main/java/org/motechproject/nms/region/domain/Taluka.java [hunk 3]
```diff
@@ -46,6 +49,10 @@
     @NotNull
     private District district;
 
+    // used by CSV import as a placeholder for the district
+    @Ignore
+    private String districtCode;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/domain/Taluka.java]: As above

File: region/src/main/java/org/motechproject/nms/region/domain/HealthSubFacility.java [hunk 4]
```diff
@@ -25,11 +27,13 @@
     private String regionalName;
 
     @Field
-    @Unique
     @Column(allowsNull = "false")
     @NotNull
     private Long code;
 
+    @Ignore
+    private long healthFacilityCode;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/domain/HealthSubFacility.java]: As above

File: region/src/main/java/org/motechproject/nms/region/domain/Village.java [hunk 5]
```diff
@@ -44,13 +45,21 @@
...
     @Field
+    @NotNull
+    @Min(0)
     // This is the village code for a non-census village.
-    private Long svid;
+    private long svid;
+
+    // used by CSV import as a placeholder for the taluka
+    @Ignore
+    private String talukaCode;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/domain/Village.java]: As above

File: region/src/main/java/org/motechproject/nms/region/service/impl/BaseVillageImportService.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        District district = (District) getParent(PARENT_DISTRICT);
+        if (district == null) {
+            throw new CsvImportException("No district provided!");
+        }
+
+        try {
+            taluka = talukaService.findByDistrictAndCode(district, village.getTalukaCode());
+        } catch (NumberFormatException e) {
+            throw new CsvImportDataException(String.format("Invalid taluka: %s",
+                    (String) village.getTalukaCode()), e);
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/service/impl/BaseVillageImportService.java]: Redundant cast.

File: region/src/main/java/org/motechproject/nms/region/service/CensusVillageImportService.java [hunk 7]
```diff
@@ -1,4 +1,7 @@
 package org.motechproject.nms.region.service;
 
+import org.motechproject.nms.region.domain.District;
+
 public interface CensusVillageImportService extends LocationDataImportService {
+    void addParent(District district);
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/service/CensusVillageImportService.java]: Aren't all those addParent(..) methods supposed to be called from LocationDataImportController too?

File: region/src/main/java/org/motechproject/nms/region/service/impl/BaseLocationImportService.java [hunk 8]
```diff
@@ -13,17 +13,33 @@
...
 public abstract class BaseLocationImportService<T> {
 
+    public static final String PARENT_STATE = "state";
+    public static final String PARENT_DISTRICT = "district";
+    public static final String PARENT_TALUKA = "taluka";
+    public static final String PARENT_HEALTH_BLOCK = "healthBlock";
+
     private Class<T> type;
     private MotechDataService<T> dataService;
+    private Map<String, Object> parents;
```
Comment (kamknitt) [inline region/src/main/java/org/motechproject/nms/region/service/impl/BaseLocationImportService.java]: Placing a property in a service makes the service stateful. We should avoid it, as it makes it prone to concurrency issues. Consider passing the map as a parameter to the importData(..) method.

**Comment by kamknitt** [comment]
The most crucial part to fix is the 'parents' property in the BaseLocationImportService class.
I've refactored the CsvInstanceImporter a bit (and opened a PR: https://github.com/motech-implementations/mim/pull/389) and added the ability to use values from already procesed cells to determine the value of consecutive cells. I thought that maybe you will find it useful while working on this PR.

**Comment by frankhuster** [comment]
Great feedback, thanks!
I'll get working on it.

File: region/src/main/java/org/motechproject/nms/region/service/impl/BaseVillageImportService.java (hunk 6 above)
Comment (koshalt) [inline region/src/main/java/org/motechproject/nms/region/service/impl/BaseVillageImportService.java]: removed cast.

**Comment by frankhuster** [comment]
Now uses @kamknitt's nice solution for accessing sibling column values in `CsvInstanceImporter`.

**Comment by kamknitt** [comment]
Looks good to me.

**Pull request merged**